appsec/src/extension/commands/request_shutdown.c

```diff
@@ -189,3 +261,355 @@ static void _pack_headers_no_cookies_map(
+        &zv, entity, entity_len, PHP_JSON_OBJECT_AS_ARRAY, MAX_DEPTH);
+    if (Z_TYPE(zv) == IS_NULL) {
+        mlog(dd_log_info, "Failed to parse JSON response body");
+        zval_ptr_dtor(&zv);
+    }
+    return zv;
+}
+
+static bool _assume_utf8(const char *ct, size_t ct_len)
+{
+    const char *psemi = memchr(ct, ';', ct_len);
+    if (!psemi) {
+        return true;
+    }
+    for (const char *end = ct + ct_len, *c = psemi + 1;
+         c < end - LSTRLEN("charset=utf-8") + 1; c++) {
+        if (tolower(*c) == 'c' && tolower(*(c + 1)) == 'h' &&
+            tolower(*(c + 2)) == 'a' && tolower(*(c + 3)) == 'r' &&
+            tolower(*(c + 4)) == 's' && tolower(*(c + 5)) == 'e' && // NOLINT
+            tolower(*(c + 6)) == 't') {                             // NOLINT
+            c += LSTRLEN("charset");
+            for (; c < end && *c == ' '; c++) {}
+            if (c < end && *c == '=') {
+                for (c++; c < end - LSTRLEN("utf-8") && *c == ' '; c++) {}
+                if (tolower(*c) == 'u' && tolower(*(c + 1)) == 't' &&
+                    tolower(*(c + 2)) == 'f' && tolower(*(c + 3)) == '-' &&
+                    tolower(*(c + 4)) == '8') {
+                    return true;
+                }
+                return false;
+            }
+            return true;
+        }
+    }
+    return true;
+}
+
+static zval _convert_xml_impl(const char *nonnull entity, size_t entity_len,
+    const char *content_type, size_t content_type_len)
+{
+    static zval null_zv = {.u1.type_info = IS_NULL};
+    zval function_name;
+    zval parser;
+    zval args[4];
+    int is_successful;
+
+    /* create XMLParser */
+    ZVAL_STRING(&function_name, "xml_parser_create");
+    is_successful = call_user_function(
+        CG(function_table), NULL, &function_name, &parser, 0, NULL);
+    zval_dtor(&function_name);
+
+#if PHP_VERSION_ID >= 80000
+#    define XML_PARSER_TYPE IS_OBJECT
+#else
+#    define XML_PARSER_TYPE IS_RESOURCE
+#endif
+    if (is_successful == FAILURE || Z_TYPE(parser) != XML_PARSER_TYPE) {
+        mlog(dd_log_debug, "Failed to create XML parser");
+        if (Z_TYPE(parser) == XML_PARSER_TYPE) {
+            zval_dtor(&parser);
+        }
+        return null_zv;
+    }
+
+    /* disable case folding */
+    zval retval;
+    ZVAL_STRING(&function_name, "xml_parser_set_option");
+    ZVAL_COPY_VALUE(&args[0], &parser);
+    ZVAL_LONG(&args[1], 1 /*PHP_XML_OPTION_CASE_FOLDING*/);
+    ZVAL_BOOL(&args[2], 0);
+    is_successful = call_user_function(
+        CG(function_table), NULL, &function_name, &retval, 3, args);
+    if (is_successful == FAILURE || Z_TYPE_P(&retval) != IS_TRUE) {
+        mlog(dd_log_debug, "Failed to set XML parser option");
+        zval_dtor(&function_name);
+        zval_dtor(&parser);
+        return null_zv;
+    }
+
+    /* skip whitespace */
+    ZVAL_LONG(&args[1], 4 /*PHP_XML_OPTION_SKIP_WHITE*/);
+    ZVAL_BOOL(&args[2], 1);
+    is_successful = call_user_function(
```

Comment:
IIRC during opcache preload,  user functions can sometimes return garbage, do you think that would be a concern here?

Reply:
I'm calling C-implemented functions.